lib/kernel/test/application_SUITE.erl

```diff
@@ -1944,6 +1946,81 @@ get_appls([_ | T], Res) ->
 get_appls([], Res) ->
     Res.
 
+%% Test set_env/1.
+set_env(Conf) when is_list(Conf) ->
+    ok = application:set_env([{appinc, [{own2, persist}, {not_in_app, persist}]},
+			      {unknown_app, [{key, persist}]}]),
+
+    %% own_env1 and own2 are set in appinc
+    undefined = application:get_env(appinc, own_env1),
+    {ok, persist} = application:get_env(appinc, own2),
+    {ok, persist} = application:get_env(appinc, not_in_app),
+    {ok, persist} = application:get_env(unknown_app, key),
+
+    ok = application:load(appinc()),
+    {ok, value1} = application:get_env(appinc, own_env1),
+    {ok, val2} = application:get_env(appinc, own2),
+    {ok, persist} = application:get_env(appinc, not_in_app),
+    {ok, persist} = application:get_env(unknown_app, key),
+
+    %% On reload, values are lost
+    ok = application:unload(appinc),
+    ok = application:load(appinc()),
+    {ok, value1} = application:get_env(appinc, own_env1),
+    {ok, val2} = application:get_env(appinc, own2),
+    undefined = application:get_env(appinc, not_in_app),
+
+    %% Clean up
+    ok = application:unload(appinc).
+
+%% Test set_env/2 with persistent true.
+set_env_persistent(Conf) when is_list(Conf) ->
+    Opts = [{persistent, true}],
+    ok = application:set_env([{appinc, [{own2, persist}, {not_in_app, persist}]},
+			      {unknown_app, [{key, persist}]}], Opts),
+
+    %% own_env1 and own2 are set in appinc
+    undefined = application:get_env(appinc, own_env1),
+    {ok, persist} = application:get_env(appinc, own2),
+    {ok, persist} = application:get_env(appinc, not_in_app),
+    {ok, persist} = application:get_env(unknown_app, key),
+
+    ok = application:load(appinc()),
+    {ok, value1} = application:get_env(appinc, own_env1),
+    {ok, persist} = application:get_env(appinc, own2),
+    {ok, persist} = application:get_env(appinc, not_in_app),
+    {ok, persist} = application:get_env(unknown_app, key),
+
+    %% On reload, values are not lost
+    ok = application:unload(appinc),
+    ok = application:load(appinc()),
+    {ok, value1} = application:get_env(appinc, own_env1),
+    {ok, persist} = application:get_env(appinc, own2),
+    {ok, persist} = application:get_env(appinc, not_in_app),
+
+    %% Clean up
+    ok = application:unload(appinc).
+
+set_env_errors(Conf) when is_list(Conf) ->
+    "application: 1; application name must be an atom" =
+	badarg_msg(fun() -> application:set_env([{1, []}]) end),
```

Comment:
Can't seem to find badarg_msg/1 anywhere. What did I miss?

Reply:
since the warning is produced via logger, it should be possible to get hold of the printouts by either plugging in an own handler (just exporting log/2), or a filter? Don't know if it would be easier than using the capture mechanism in ct though... just an idea.

Reply:
I was not able to get the warns through capture though. Does the `ct` mechanism ask the logger to flush? Or is the new log always sync? Otherwise  we may have races.

Reply:
No, it doesn't, and I don't think there is any way to do that since the mechanism just tells the group leader to send all io_request data asynchronously on to a given pid (the test case process). I see that I have used a 100 ms timer for the same purpose in logger_std_h_SUITE :-()

Reply:
If it helps, in Elixir we have both capture_log and capture_io. capture_log is a convenience  that  adds a handler to the logger and flushes it before removing the handler. Maybe `ct` could provide something similar?

In any case, I have added assertions that invoke duplicate apps and duplicate keys with a note that it will raise but without asserting on the warnings. Once we make it raise, we can assert on them precisely. Or do you prefer to use capture with a timer?

Reply:
Yes, capture_log is a good idea, I think I'll write a ticket to add that. 

Using capture with a timer for now would be good, I think. Thanks!

Reply:
capture with a timer has been added.